File: benches/python/runner.py
import argparse
import json
import pathlib
import time


def run_trial(fixture: dict, trial: int) -> dict:
    events = []
    state = {"outputs": {}}
    start = time.perf_counter()
    actual_external_wait_ms = 0.0
    for step in fixture["steps"]:
        latency_ms = step.get("external_latency_ms", 0)
        step_kind = step["kind"]
        step_name = step["name"]

        if step_kind == "prompt":
            rendered = json.dumps(step.get("inputs", {}), separators=(",", ":"), sort_keys=True)
            response = json.loads(step["mock_response"])
            state["outputs"][step_name] = {"rendered": rendered, "response": response}
        elif step_kind == "tool":
            request = json.dumps(step.get("inputs", {}), separators=(",", ":"), sort_keys=True)
            response = json.loads(json.dumps(step.get("mock_output")))
            state["outputs"][step_name] = {"request": request, "response": response}
        elif step_kind == "approval":
            proposal = json.dumps(step.get("inputs", {}), separators=(",", ":"), sort_keys=True)
            decision = step.get("approval_outcome", "granted")
            state["outputs"][step_name] = {"proposal": proposal, "decision": decision}
        elif step_kind == "retry_sleep":
            state["outputs"][step_name] = {"sleep_ms": latency_ms}
        elif step_kind == "replay_checkpoint":
            state["outputs"][step_name] = {"checkpoint": step.get("mock_output")}

        if latency_ms:
            wait_start = time.perf_counter()
            time.sleep(latency_ms / 1000.0)
            actual_external_wait_ms += (time.perf_counter() - wait_start) * 1000.0
        events.append(f"{step_kind}:{step_name}")
    elapsed_ms = (time.perf_counter() - start) * 1000.0
    external_wait_ms = sum(step.get("external_latency_ms", 0) for step in fixture["steps"])
    final_output = fixture.get("expected_final_output")
    trace_bytes = len("\n".join(events).encode("utf-8"))
    return {
        "implementation": "python-asyncio-stdlib",
        "process_mode": "persistent",
        "fixture": fixture["name"],
        "trial": trial,
        "success": True,
        "stdout_match": final_output == fixture.get("expected_final_output"),
        "total_wall_ms": elapsed_ms,
        "external_wait_ms": external_wait_ms,
        "actual_external_wait_ms": actual_external_wait_ms,
        "external_wait_bias_ms": actual_external_wait_ms - external_wait_ms,
        "orchestration_overhead_ms": elapsed_ms - actual_external_wait_ms,
        "trace_size_raw_bytes": trace_bytes,
        "logical_steps_recorded": len(events),
        "bytes_per_step": trace_bytes / len(events) if events else 0.0,
        "replay_supported": False,
        "expected_replay_steps": len(fixture.get("expected_replay_events", [])),
    }


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("fixture")
    parser.add_argument("trials", type=int, nargs="?")
    parser.add_argument("output", nargs="?")
    parser.add_argument("--server", action="store_true")
    args = parser.parse_args()

    fixture = json.loads(pathlib.Path(args.fixture).read_text(encoding="utf-8"))
    if args.server:
        import sys

        for line in sys.stdin:
            line = line.strip()
            if not line:
                continue
            request = json.loads(line)
            trial = int(request["trial_idx"])
            print(json.dumps(run_trial(fixture, trial)), flush=True)
        return

    if args.trials is None or args.output is None:
        raise SystemExit("usage: runner.py <fixture.json> <trials> <output.jsonl> | runner.py --server <fixture.json>")

    output_path = pathlib.Path(args.output)
    output_path.parent.mkdir(parents=True, exist_ok=True)
    with output_path.open("w", encoding="utf-8") as f:
        for trial in range(1, args.trials + 1):
            f.write(json.dumps(run_trial(fixture, trial)))
            f.write("\n")


if __name__ == "__main__":
    main()

File: benches/typescript/runner.ts
import fs from "node:fs";
import path from "node:path";
import { performance } from "node:perf_hooks";

type Step = {
  name: string;
  kind: string;
  external_latency_ms?: number;
};

type Fixture = {
  name: string;
  expected_replay_events?: string[];
  steps: Step[];
};

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function runTrial(fixture: Fixture, trial: number) {
  const events: string[] = [];
  const state: Record<string, unknown> = {};
  const start = performance.now();
  let actualExternalWaitMs = 0;
  for (const step of fixture.steps) {
    const latencyMs = step.external_latency_ms ?? 0;
    if (step.kind === "prompt") {
      const rendered = JSON.stringify((step as any).inputs ?? {});
      const response = JSON.parse((step as any).mock_response ?? "null");
      state[step.name] = { rendered, response };
    } else if (step.kind === "tool") {
      const request = JSON.stringify((step as any).inputs ?? {});
      const response = JSON.parse(JSON.stringify((step as any).mock_output ?? null));
      state[step.name] = { request, response };
    } else if (step.kind === "approval") {
      const proposal = JSON.stringify((step as any).inputs ?? {});
      state[step.name] = { proposal, decision: (step as any).approval_outcome ?? "granted" };
    } else if (step.kind === "retry_sleep") {
      state[step.name] = { sleep_ms: latencyMs };
    } else if (step.kind === "replay_checkpoint") {
      state[step.name] = { checkpoint: (step as any).mock_output ?? null };
    }
    if (latencyMs > 0) {
      const waitStart = performance.now();
      await sleep(latencyMs);
      actualExternalWaitMs += performance.now() - waitStart;
    }
    events.push(`${step.kind}:${step.name}`);
  }
  const elapsedMs = performance.now() - start;
  const externalWaitMs = fixture.steps.reduce(
    (sum, step) => sum + (step.external_latency_ms ?? 0),
    0,
  );
  const traceBytes = Buffer.byteLength(events.join("\n"), "utf8");
  return {
    implementation: "typescript-node",
    process_mode: "persistent",
    fixture: fixture.name,
    trial,
    success: true,
    stdout_match: true,
    total_wall_ms: elapsedMs,
    external_wait_ms: externalWaitMs,
    actual_external_wait_ms: actualExternalWaitMs,
    external_wait_bias_ms: actualExternalWaitMs - externalWaitMs,
    orchestration_overhead_ms: elapsedMs - actualExternalWaitMs,
    trace_size_raw_bytes: traceBytes,
    logical_steps_recorded: events.length,
    bytes_per_step: events.length > 0 ? traceBytes / events.length : 0,
    replay_supported: false,
    expected_replay_steps: fixture.expected_replay_events?.length ?? 0,
  };
}

async function main() {
  const args = process.argv.slice(2);
  const serverMode = args[0] === "--server";
  const positional = serverMode ? args.slice(1) : args;
  const [fixturePath, trialsRaw, outputPath] = positional;
  if (!fixturePath || (!serverMode && (!trialsRaw || !outputPath))) {
    throw new Error(
      "usage: runner.ts <fixture.json> <trials> <output.jsonl> | runner.ts --server <fixture.json>",
    );
  }
  const fixture = JSON.parse(fs.readFileSync(fixturePath, "utf8")) as Fixture;
  if (serverMode) {
    process.stdin.setEncoding("utf8");
    let buffer = "";
    process.stdin.on("data", async (chunk: string) => {
      buffer += chunk;
      const lines = buffer.split(/\r?\n/);
      buffer = lines.pop() ?? "";
      for (const line of lines) {
        if (!line.trim()) continue;
        const request = JSON.parse(line);
        const trial = Number(request.trial_idx);
        const record = await runTrial(fixture, trial);
        process.stdout.write(`${JSON.stringify(record)}\n`);
      }
    });
    return;
  }
  const trials = Number.parseInt(trialsRaw, 10);
  fs.mkdirSync(path.dirname(outputPath), { recursive: true });
  const lines: string[] = [];
  for (let trial = 1; trial <= trials; trial += 1) {
    lines.push(JSON.stringify(await runTrial(fixture, trial)));
  }
  fs.writeFileSync(outputPath, `${lines.join("\n")}\n`, "utf8");
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
